thoth/storages/graph/postgres.py

```diff
@@ -3775,6 +3775,98 @@ def create_python_cve_record(
+        package_name: str,
+        package_version: str,
+        index_url: str,
+        value: bool,
+    ) -> None:
+        """Update value of is_missing flag for PythonPackageVersion."""
+        with self._session_scope() as session:
+            subq = (session.query(PythonPackageVersion)
+                    .join(PythonPackageIndex)
+                    .filter(PythonPackageVersion.package_name == package_name)
+                    .filter(PythonPackageVersion.package_version == package_version)
+                    .filter(PythonPackageIndex.url == index_url)
+                    .with_entities(PythonPackageVersion.id)
+                    )
+            (
+                session.query(PythonPackageVersion)
+                .filter(PythonPackageVersion.id.in_(subq))
+                .update({"is_missing": value}, synchronize_session='fetch')
+            )
+
+    def get_all_repositories_using_package(self, package_name: str, index_url: str) -> List[str]:
+        """Given package info, get unique urls of repos which have been advised with this package."""
+        with self._session_scope() as session:
+            result = (
+                session.query(PythonPackageVersion)
+                .join(PythonPackageIndex)
+                .join(PythonRequirementsLock)
+                .join(AdviserRun, AdviserRun.user_software_stack_id == PythonRequirementsLock.python_software_stack_id)
+                .order_by(AdviserRun.origin)
+                .order_by(AdviserRun.datetime.desc())
+                .distinct(AdviserRun.origin)
+                .filter(PythonPackageIndex.url == index_url)
+                .filter(PythonPackageVersion.package_name == package_name)
+                .with_entities(AdviserRun.origin)
+            ).all()
+            result = [x[0] for x in result]
+            return result
+
+    def get_all_repositories_using_package_version(
+        self,
+        package_name: str,
+        package_version: str,
+        index_url: str,
+    ) -> List[str]:
+        """Given package version info, get unique urls of repos which have been advised with this package version."""
+        with self._session_scope() as session:
+            result = (
+                session.query(PythonPackageVersion)
+                .join(PythonPackageIndex)
+                .join(PythonRequirementsLock)
+                .join(AdviserRun, AdviserRun.user_software_stack_id == PythonRequirementsLock.python_software_stack_id)
+                .order_by(AdviserRun.origin)
+                .order_by(AdviserRun.datetime.desc())
+                .distinct(AdviserRun.origin)
+                .filter(PythonPackageIndex.url == index_url)
+                .filter(PythonPackageVersion.package_name == package_name)
+                .filter(PythonPackageVersion.package_version == package_version)
+                .with_entities(AdviserRun.origin)
+            ).all()
+            result = [x[0] for x in result]
+            return result
```

Comment:
This query is redundant and can be merged with the one above, adding an optional filter for the version if present :)